### PR TITLE
Implementation of discussed changes

### PR DESCRIPTION
**Pull request by vidartf**
This refactors widgets, and should implement most of the discussed changes of the ROI. Some needed cross-contamination with NEW_Events and NEW_Interactive.

### CODE REVIEW

**Comment by francisco-dlp** [comment]
This is awesome, I thought that it'll take us weeks to get all of this done and you've done it in days (!). I bet that you have worked in multiple branches and merged them all into NEW_ROI, could you pull request to the individual branches too? In this way we'll keep things tidy what should make it easier for others to  discuss and review your work.

**Comment by vidartf** [comment]
I mostly worked in one branch, but I'll cherry-pick the commits out to the
corresponding branches and PR those.

**Comment by vidartf** [comment]
Also added 1D ROI with appropriate widget.

File: hyperspy/drawing/widgets.py [hunk 1]
```diff
@@ -67,40 +71,48 @@ def set_on(self, value):
...
 
-    def add_patch_to(self, ax):
-        self.set_patch()
+    def _add_patch_to(self, ax):
+        self._set_patch()
         ax.add_artist(self.patch)
         self.patch.set_animated(hasattr(ax, 'hspy_fig'))
 
-    def add_axes(self, ax):
+    def set_axis(self, ax):
```
Comment (francisco-dlp) [inline hyperspy/drawing/widgets.py]: I agree that `add_axes` was misleading. However, `ax` should be a MPL Axes instance, so `set_axis` doesn't feel fully right either. What about `set_mpl_axes` or `set_mpl_ax`?

**Comment by francisco-dlp** [comment]
Nice.

As previously discussed, this branch has a massive amount of changes and it would have been better to perform many of these changes in separate branches. For example, here, in addition to implementing the ROIs, you have fixed bugs, renamed attributes, enhanced some features and added others... All in all, it is an amazing contribution. However, it is difficult to review and to discuss the changes when they are all coming in one PR. If you had made several PRs many of them would be already merged, however, as it is, we have to review and agree on the full set. Beside, renaming attributes is often dangerous, specially in the drawing code that does not yet benefit from unittests. So, renaming existing attributes should be done in isolated branches and we should always make sure that we perform the renaming in all files.

**Comment by vidartf** [comment]
Agree on most of your points, I'll just have to figure out the hierarchy of which changes that depend on each other. Do you think I should push here, or to hyperspy master?

**Comment by francisco-dlp** [comment]
It depends, the commits that deal only the new ROIs should come here, but everything else should go to hyperspy/master.

Sorry that we don't have a good Guide for developers yet, I just opened #445 to fix this.

**Comment by vidartf** [comment]
Well, I tried to move some stuff out, but most of what happens in widget.py and roi.py is interdependent, especially with the big revamp of widget.py that was needed to rewire everything. I'm not sure I can break it down into meaningful chunks that won't have a broken intermediate state.

**Comment by vidartf** [comment]
As an example, in 382de116e36de4a91b6e0b5cf47a7e707938cae4 I moved _a lot _ of code around to extract common tasks higher in widget class hierarchy, which of source makes the diff a mess. And the state of the code before that was rather WIP. If all those changes have to broken down into separate items, each with a workable code base, then I simply won't have the time necessary. To be honest, I have lost oversight over all the different pull requests, and which branches have which commits etc. Hopefully that doesn't mean any bugs will slip through, but the odds certainly have gone up.

On the other hand, if I could help code review by adding docstrings then that's something I'd be happy to do.

**Comment by francisco-dlp** [comment]
The situation has vastly improved, thanks!

**Comment by francisco-dlp** [comment]
Could you merge master into your "NEW_ROI" branch? It is not straightforward and you're in a better position than I to perform the task without breaking something...

**Comment by vidartf** [comment]
That should do it, right?

**Comment by vidartf** [comment]
No, I still need to fix some stuff, let me do it quickly.

**Comment by vidartf** [comment]
There, I also merged your branch so that it can auto-merge,

**Comment by vidartf** [comment]
Now I also removed the code relevant to discussion here francisco-dlp/hyperspy#6. Probably better to have that as separate PR :)

Now I think you can merge, if you approve so far. I'll try to doc up the code as soon as I have time, but there will be a bit of a crunch for me going forward.

**Comment by vidartf** [comment]
I'm trying to separate the pure ROI code and widget code now. This PR should now only be ROI code. The diff will be a lot simpler if you merge your NEW_Events first.

**Comment by vidartf** [comment]
For now, I've put all the widget code here:
https://github.com/vidartf/hyperspy/tree/NEW_widgets2

**Comment by vidartf** [comment]
Finally, I've put the code that connects the new ROIs with the new widgets here:
https://github.com/vidartf/hyperspy/tree/NEW_ROI_interactive3

**Comment by dnjohnstone** [comment]
For some things I'm working on at the moment I have a stack of images and I want to plot the intensity in a subset of pixels for each image as an array with dimensions equal to my navigation dimensions - I understand that this is essentially done somewhere in this PR and all I'd really like to do is extend the shapes of the subset of pixels that are being plotted e.g. a circle around a point, an annulus around the centre...BUT I don't want to make this branch any more complicated than it already is so @vidartf I was wondering if you've got plans to separate things out any further?

**Comment by vidartf** [comment]
@dnjohnstone I'm afraid my earlier attempt at separating things is horribly dated (and quite possibly broken) due to the branch hell it led to. I still have workable "higher level" branches though, so I should be able to extract a branch for what you need. From how you describe it, and what I know of your work, you don't need any widgets to manually interact with the ROIs? If so, I can just isolate the code for the ROIs. I have a circular ROI currently, but not ellipsis yet. Let me know if you need that and I'll rush it.

Technical note: Currently there is no support for "soft" edges, i.e. the pixels/voxels are either selected or not. Adding feathering is of course possible, but I/we would have to figure out how to best implement it (partial masking / alpha channel would be needed).

**Comment by vidartf** [comment]
I think a decent ROI-only implementation is now in [vidartf/hyperspy/NEW_ROI_only](https://github.com/vidartf/hyperspy/tree/NEW_ROI_only).

**Comment by dnjohnstone** [comment]
Ok thanks for that! That's exactly right, I only need the ROI stuff for now so that's perfect. I'll have a look in detail soon.

Regarding soft edges: This is an interesting point for masking when doing decompositions as well. By coincidence I was just thinking of looking in to smooth edges for masks because I'm using a circle there and it's just ugly without soft edges - would probably be neat if we can settle on the best way to do this and use it generally.

**Pull request closed without merging**